pkg/fqdn/dnsproxy/udp.go

```diff
@@ -30,14 +31,48 @@ var udpOOBSize = func() int {
 	return int(unsafe.Sizeof(hdr) + unsafe.Sizeof(addr))
 }()
 
+// Set up new SessionUDPFactory with dedicated raw socket for sending responses.
+//   - Must use a raw UDP socket for sending responses so that we can send
+//     from a specific port without binding to it.
+//   - The raw UDP socket must be bound to a specific IP address to prevent
+//     it receiving ALL UDP packets on the host.
+//   - We use oob data to override the source IP address when sending
+//   - Must use separate sockets for IPv4/IPv6, as sending to a v6-mapped
+//     v4 address from a socket bound to "::1" does not work due to kernel
+//     checking that a route exists from the source address before
+//     the source address is replaced with the (transparently) changed one
+func NewSessionUDPFactory(ipFamily ipfamily.IPFamily) (dns.SessionUDPFactory, error) {
+	var err error
+	var rawconn4 *net.IPConn
+	var rawconn6 *net.IPConn
+
+	if ipFamily.IPv4Enabled {
+		rawconn4, err = bindUDP("127.0.0.1", true, false) // raw socket for sending IPv4
+		if err != nil {
+			return nil, fmt.Errorf("failed to open raw UDP IPv4 socket for DNS Proxy: %w", err)
+		}
+	}
+
+	if ipFamily.IPv6Enabled {
+		rawconn6, err = bindUDP("::1", false, true) // raw socket for sending IPv6
+		if err != nil {
+			return nil, fmt.Errorf("failed to open raw UDP IPv6 socket for DNS Proxy: %w", err)
+		}
+	}
+
+	return &sessionUDPFactory{
+		responseConn4: rawconn4,
+		responseConn6: rawconn6,
+	}, nil
+}
+
 type sessionUDPFactory struct {
 	// A pool for UDP message buffers.
 	udpPool sync.Pool
 
-	// ipv4Enabled and ipv6Enabled are used when setting up the proxy sockets
-	// later, and determine if we bind to 127.0.0.1 and ::1, respectively.
-	// See sessionUDPFactory.SetSocketOptions
-	ipv4Enabled, ipv6Enabled bool
+	// responseConn4 and responseConn6 are used to send the response
+	// See sessionUDP.WriteResponse
+	responseConn4, responseConn6 *net.IPConn
```

Comment:
How about keeping only one pointer (`responseConn`) in the factory? Each factory instance only has one family defined, so we do not need two pointers here?

Reply:
The reason is that the "building blocks" (e.g. SessionUDPFactory, setting transparent socket options, ...) were all built when a socket bound to a Cilium DNS server supported multiple IP Famliies (IPv4 & IPv6).

But this is no longer necessary since we use a dedicated DNS Server (& socket) per IPFamily.

Therefore i agree that this could simplified.

=> I refactored the SessionUDPFactory & "setting the transparent socket options" to only support one IPFamily. All the relevant information are read from the `IPFamily` struct that gets passed down - to keep the code agnostic of the actual IP Family.